core-primitives/stf-executor/src/executor_tests.rs

```diff
@@ -0,0 +1,317 @@
+	let call_operation_hash: H256 =
+		blake2_256(&signed_call.clone().into_trusted_operation(true).encode()).into();
+	let signed_call_two = TrustedCall::balance_transfer(
+		sender.public().into(),
+		sender.public().into(),
+		100,
+	)
+	.sign(&sender.clone().into(), 0, &mrenclave, &shard);
+
+	let old_state_hash = state_hash(&state_handler.load_initialized(&shard).unwrap());
+
+	// when
+	let batch_execution_result = stf_executor
+		.propose_state_update(
+			&vec![signed_call.clone(), signed_call_two.clone()],
+			&ParentchainHeaderBuilder::default().build(),
+			&shard,
+			Duration::ZERO,
+			|state| state,
+		)
+		.unwrap();
+
+	// then
+	assert_eq!(old_state_hash, batch_execution_result.state_hash_before_execution);
+	assert_eq!(batch_execution_result.executed_operations.len(), 1);
+	assert_eq!(batch_execution_result.get_executed_operation_hashes(), vec![call_operation_hash]);
+	// Ensure that state has been updated and not actually written.
+	assert_ne!(
+		state_handler.load_initialized(&shard).unwrap(),
+		batch_execution_result.state_after_execution
+	);
+}
+
+pub fn propose_state_update_always_executes_preprocessing_step() {
+	// given
+	let shard = ShardIdentifier::default();
+	let (stf_executor, _, state_handler) = stf_executor();
+	let key = "my_key".encode();
+	let value = "my_value".encode();
+	let old_state_hash = state_hash(&state_handler.load_initialized(&shard).unwrap());
+
+	// when
+	let batch_execution_result = stf_executor
+		.propose_state_update(
+			&vec![],
+			&ParentchainHeaderBuilder::default().build(),
+			&shard,
+			Duration::ZERO,
+			|mut state| {
+				state.insert(key.clone(), value.clone());
+				state
+			},
+		)
+		.unwrap();
+
+	// then
+	assert_eq!(old_state_hash, batch_execution_result.state_hash_before_execution);
+
+	// Ensure that state has been updated.
+	let old_state = state_handler.load_initialized(&shard).unwrap();
+	let retrieved_value = batch_execution_result.state_after_execution.get(key.as_slice()).unwrap();
+	assert_eq!(*retrieved_value, value);
+	// Ensure that state has not been actually written.
+	assert_ne!(old_state, batch_execution_result.state_after_execution);
+}
+
+pub fn execute_timed_getters_batch_executes_if_enough_time() {
+	// given
+	let sender = endowed_account();
+	let trusted_getter = TrustedGetter::free_balance(sender.public().into()).sign(&sender.into());
+	let (stf_executor, _, state_handler) = stf_executor();
+	let (_, shard) = init_state_and_shard_with_state_handler(state_handler.as_ref());
+	let execution_duration = Duration::from_secs(10000);
+
+	// when
+	assert!(panic::catch_unwind(|| {
+		stf_executor.execute_timed_getters_batch(
+			&vec![trusted_getter.clone()],
+			&shard,
+			execution_duration,
+			|trusted_getter_signed: &TrustedGetterSigned,
+			 _state_result: Result<Option<Vec<u8>>>| {
+				// then
+				assert_eq!(*trusted_getter_signed, trusted_getter);
+				panic!("test should enter here");
+			},
+		)
+	})
+	.is_err());
```

Comment:
Why do we test it like that and not simply wait until the time's up?

Reply:
In this test I want to ensure that the function we're giving as an input is entered (i.e. it must panic). Simply waiting for the time to be up does not ascertain us that we did not forget to call the input function in the implementation.

Reply:
Ah, or do you mean stalling time in the function and then check for the time it took to execute? Sure, that'd work too, but it would take longer, than simply panicking, I guess.